source/Dictionary.cpp

```diff
@@ -44,48 +47,96 @@ namespace {
 		return make_pair(low, false);
 	}
 
-	// String interning: return a pointer to a character string that matches the
-	// given string but has static storage duration.
-	const char *Intern(const char *key)
+	// Creates new if missing.
+	static dict NameToKey(const string &name)
 	{
-		static set<string> interned;
 		static mutex m;
 
-		// Just in case this function is accessed from multiple threads:
 		lock_guard<mutex> lock(m);
-		return interned.insert(key).first->c_str();
+		map<string, dict>::iterator it = nameToKey.lower_bound(name);
+		if(it == nameToKey.end() || nameToKey.key_comp()(name, it->first))
+		{
+			it = nameToKey.insert(it, pair<string, dict>(name, names.size()));
+			names.push_back(it->first.c_str());
+		}
+		return it->second;
 	}
 }
 
 
 
-double &Dictionary::operator[](const char *key)
+double &Dictionary::operator[](const char *name)
+{
+	pair<size_t, bool> pos = Search(name, *this);
+	if(pos.second)
+		return store.data()[data()[pos.first]];
```

Comment:
There is no need to call `data()` before indexing; `std::vector` has `operator[]` which does this automatically. This also applies to all other uses of `data()` in this PR.